### PR TITLE
Merging Fully Implemented Group Reservations

### PR DESCRIPTION
**Pull request by tmattvu**
- As a student, when making a room reservation, I now have to add users to my reservation in order to meet the capacity that each room requires.

### CODE REVIEW

File: frontend/src/app/coworking/coworking.models.ts [hunk 1]
```diff
@@ -38,6 +38,17 @@ export interface Room {
...
+  id: string | null;
+  nickname: string;
+  building: string;
+  room: string;
+  capacity: number;
+  reservable: boolean;
+  description: string;
+  seats: Seat[];
+}
+
```
Comment (ajaygandecha) [inline frontend/src/app/coworking/coworking.models.ts:51]: Typically, the frontend does not use `Details` models because dealing with types becomes more difficult - we instead have used `| null` on all of the fields that would be part of `Details` not part of the original model.

If we must use a details model here, make it extend `Room`!

File: frontend/src/app/coworking/room-reservation/reservation-table.service.ts [hunk 2]
```diff
@@ -85,19 +96,47 @@ export class ReservationTableService {
     );
   }
 
+  getRoomInformation(): Observable<RoomDetails[]> {
+    return this.http.get<RoomDetails[]>(`/api/coworking/rooms/`);
+  }
```
Comment (ajaygandecha) [inline frontend/src/app/coworking/room-reservation/reservation-table.service.ts:101]: Instead of adding it to this service, it would be better to inject the academics service, which has a call for using rooms. Better yet, since rooms is being used in many places, it would probably be more ideal for us to create a completely separate `RoomService` that is injected into all these places - perhaps we can make a ticket for this.

File: frontend/src/app/coworking/room-reservation/reservation-table.service.ts [hunk 3]
```diff
@@ -85,19 +96,47 @@ export class ReservationTableService {
...
+            return reservationRequest;
+          }),
+          switchMap((reservationRequest) =>
+            this.makeDraftReservation(reservationRequest)
+          ),
+          catchError((err) =>
+            throwError(() => new Error('Not enough people to reserve!'))
+          )
+        );
+      })
```
Comment (ajaygandecha) [inline frontend/src/app/coworking/room-reservation/reservation-table.service.ts:138]: This is a lot of logic to have in a frontend service - it might be better to move this into a backend service connected to an API. There can be errors that are resurfaced from the API if rooms do not exist or if there are not enough users. Thoughts @KrisJordan?

File: frontend/src/app/coworking/room-reservation/reservation-table.service.ts [hunk 4]
```diff
@@ -110,6 +149,18 @@ export class ReservationTableService {
     );
   }
 
+  publicToProfile(users: (PublicProfile | Profile)[]): Observable<Profile[]> {
+    //Converts publicProfiles to Profiles to send into backend
+    const userRequests = users.map((user) =>
+      this.http.get<Profile>(`/api/profile/users/${user.id}`)
+    );
+
+    return forkJoin(userRequests).pipe(
```
Comment (ajaygandecha) [inline frontend/src/app/coworking/room-reservation/reservation-table.service.ts:158]: Is there a reason we need `forkJoin` here? Might need some documentation.

File: frontend/src/app/coworking/widgets/room-dialogue/room-dialogue.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (ajaygandecha) [inline frontend/src/app/coworking/widgets/room-dialogue/room-dialogue.component.spec.ts:1]: Spec file can be removed

File: backend/api/coworking/reservation.py [hunk 6]
```diff
@@ -105,3 +106,10 @@ def get_total_hours_study_room_reservations(
 ) -> str:
     """Allows a user to know how many hours they have reserved in all study rooms (Excludes CSXL)."""
     return reservation_svc._get_total_time_user_reservations(subject)
+
+
+@api.get("/rooms/", tags=["Coworking"])
```
Comment (ajaygandecha) [inline backend/api/coworking/reservation.py:111]: Interesting question (will loop in @KrisJordan), do we want this in the rooms API file instead?

File: frontend/src/app/coworking/widgets/room-reservation-table/room-reservation-table.widget.html [hunk 7]
```diff
@@ -26,4 +38,11 @@
     [disabled]="!selectButton">
     Reserve
   </button>
+
+  <button
+    mat-stroked-button
+    class="button"
+    (click)="openMemberSelectionDialog()">
+    Add Users
+  </button>
```
Comment (ajaygandecha) [inline frontend/src/app/coworking/widgets/room-reservation-table/room-reservation-table.widget.html:47]: Looping @KrisJordan in - I might argue that moving the selector dropdown right to this page directly might be better from a UI flow perspective, esp on mobile.

For UI purposes, it is also possible to to add a permanent chip for the current user into the selector. So, we can ultimately just one selector field with a permanent "current user" chip, then the ability to add more members. This selector can appear once a room has been selected on the table above (we can do this with `*ngIf`)

If you have questions, feel free to reply here!

File: frontend/src/app/coworking/widgets/room-reservation-table/room-reservation-table.widget.ts [hunk 8]
```diff
@@ -148,9 +182,14 @@ export class RoomReservationWidgetComponent {
       },
 
       (error) => {
-        // Handle errors here
-        console.error('Error drafting reservation', error);
-        this.snackBar.open(error.error.message, 'Close', this.snackBarOptions);
+        let errorMessage = 'Unknown error';
+        try {
+          const errorObj = JSON.parse(error.message);
+          errorMessage = errorObj.error.message;
```
Comment (ajaygandecha) [inline frontend/src/app/coworking/widgets/room-reservation-table/room-reservation-table.widget.ts:188]: Nit: Is it possible just to condense this?

File: frontend/src/app/coworking/widgets/room-dialogue/room-dialogue.component.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        class="dialog-close-button"
+        mat-icon-button
+        (click)="onCloseClick()">
+        <mat-icon>close</mat-icon>
+      </button>
+      <h1 class="dialog-title" mat-dialog-title>Room Information</h1>
+      <div class="dialog-content" mat-dialog-content>
+        <p>Room: {{ data.id }}</p>
+        <p>Capacity: {{ data.capacity }}</p>
+        <p>Description: {{ data.description }}</p>
```
Comment (ajaygandecha) [inline frontend/src/app/coworking/widgets/room-dialogue/room-dialogue.component.ts:18]: Might be a good idea to either provide a placeholder, or hide this entirely, if a description for the room is blank

File: frontend/src/app/coworking/widgets/group-reservation-card/group-reservation-card.widget.html [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    style="
+      display: flex;
+      justify-content: space-between;
+      margin: 12px 8px 45px;
+      box-sizing: border-box;
+    ">
+    <!-- Add Users Button -->
+    <button
+      mat-button
+      color="primary"
```
Comment (ajaygandecha) [inline frontend/src/app/coworking/widgets/group-reservation-card/group-reservation-card.widget.html:37]: Use the `outlined` style for buttons! Also, these buttons can be bottom-aligned

File: frontend/src/app/coworking/room-reservation/reservation-table.service.ts [hunk 11]
```diff
@@ -85,19 +96,47 @@ export class ReservationTableService {
...
-      operationStart
+        return this.getRoomInformation().pipe(
+          map((rooms) => {
+            const room = rooms.find((room) => room.id === selectedRoom.room);
+            if (!room) {
+              throw new Error('Room not found');
+            }
+            if (users.length < room.capacity) {
+              throw new Error('Not enough people to reserve');
+            }
```
Comment (ajaygandecha) [inline frontend/src/app/coworking/room-reservation/reservation-table.service.ts:123]: Did you check for max capacity? On localhost, I tried reserving a room w/ capacity 2 with 3 users, and it succeeded.

File: frontend/src/app/academics/academics-admin/room/room-editor/room-editor.component.html [hunk 12]
```diff
@@ -65,6 +65,15 @@
...
           required />
       </mat-form-field>
+      <mat-form-field appearance="outline" color="accent">
+        <mat-label>Room Description</mat-label>
+        <input
+          matInput
+          placeholder=""
+          formControlName="description"
+          name="description"
+          required />
```
Comment (ajaygandecha) [inline frontend/src/app/academics/academics-admin/room/room-editor/room-editor.component.html:75]: Thoughts on removing `required` from this field? - since description is a new field (and all rooms do not have a description by default), going to edit any room will now force the user to add a description if they want to save. Could always default to an empty string.

File: backend/models/room_details.py [hunk 13]
```diff
@@ -24,3 +25,6 @@ def to_room(self) -> Room:
         Returns:
             Room: The model representation of the entity."""
         return Room(id=self.id, nickname=self.nickname)
+
+    def __hash__(self):
+        return hash(self.id)
```
Comment (ajaygandecha) [inline backend/models/room_details.py:30]: What is this for?

**Review by ajaygandecha — CHANGES_REQUESTED** [review]
Great start on group reservations @tmattvu! I left a few comments on some backend and frontend fixes. 

Also, it does not seem this PR includes any updates to the backend tests for room reservations - two tests fail and overall reservation test coverage has dropped to 86%. Please make sure to test the changes you have added here for different edge cases (found the upper bound capacity issue when on localhost!) and commit that here!

File: frontend/src/app/coworking/room-reservation/reservation-table.service.ts (hunk 11 above)
Comment (tmattvu) [reply frontend/src/app/coworking/room-reservation/reservation-table.service.ts:123]: Hey Ajay, just skimming the CR's before I tackle them tomorrow,but for this one, I didn't check for more than the capacity because for example, if a group of 3 wants to use one of the pair programming rooms, I assumed that'd be fine. But if we want it to where it has to be exactly the capacity, I can change that easily!

Comment (ajaygandecha) [reply frontend/src/app/coworking/room-reservation/reservation-table.service.ts:123]: Sounds good, can loop in @KrisJordan for this one!

File: backend/api/profile.py [hunk 14]
```diff
@@ -40,6 +40,24 @@ def read_profile(
...
+):
+    """
+    Retrieve a user's profile by ID.
+    If the user does not exist, return a placeholder for unregistered user
+    """
+    user = user_svc.get_by_id(user_id)
+    if user:
+        return user
+    else:
+        return UnregisteredUser(id=user_id)
```
[the remark this answers is not in the document]
Comment (tmattvu) [reply backend/api/profile.py:58]: Fixed!

File: frontend/src/app/coworking/room-reservation/reservation-table.service.ts (hunk 3 above)
Comment (tmattvu) [reply frontend/src/app/coworking/room-reservation/reservation-table.service.ts:138]: DraftReservation has been refactored!

File: frontend/src/app/coworking/widgets/room-reservation-table/room-reservation-table.widget.html (hunk 7 above)
Comment (tmattvu) [reply frontend/src/app/coworking/widgets/room-reservation-table/room-reservation-table.widget.html:47]: Left as is for now since Kris has not answered yet!

File: frontend/src/app/coworking/widgets/group-reservation-card/group-reservation-card.widget.html (hunk 10 above)
Comment (tmattvu) [reply frontend/src/app/coworking/widgets/group-reservation-card/group-reservation-card.widget.html:37]: Changed to a mat-stroked-button and bottom aligned the buttons to the bottom of the dialog box. However, I'd argue that having the buttons closer to the selected members section looks better (if this is what you meant by bottom align)! 

Snippets:

Without bottom aligning (the box has to be this big to prevent overflow when selecting multiple members):
image

With bottom aligning:
image

**Comment by tmattvu** [comment]
Fixed CR's relating to my code (group reservations). Aarjav will fix the rest of the CR's relating more towards room reservations and the new code for descriptions and capacity!

File: backend/models/room_details.py (hunk 13 above)
Comment (tmattvu) [reply backend/models/room_details.py:30]: Looping in @aarjavjain2002

File: frontend/src/app/academics/academics-admin/room/room-editor/room-editor.component.html (hunk 12 above)
Comment (tmattvu) [reply frontend/src/app/academics/academics-admin/room/room-editor/room-editor.component.html:75]: Looping in @aarjavjain2002

File: frontend/src/app/coworking/widgets/room-dialogue/room-dialogue.component.ts (hunk 9 above)
Comment (tmattvu) [reply frontend/src/app/coworking/widgets/room-dialogue/room-dialogue.component.ts:18]: Looping in @aarjavjain2002

File: frontend/src/app/coworking/widgets/room-reservation-table/room-reservation-table.widget.ts (hunk 8 above)
Comment (tmattvu) [reply frontend/src/app/coworking/widgets/room-reservation-table/room-reservation-table.widget.ts:199]: Looping in @aarjavjain2002

File: backend/api/coworking/reservation.py (hunk 6 above)
Comment (tmattvu) [reply backend/api/coworking/reservation.py:111]: Left as is since Kris has not answered! Also looping in @aarjavjain2002

File: frontend/src/app/coworking/widgets/room-dialogue/room-dialogue.component.spec.ts (hunk 5 above)
Comment (tmattvu) [reply frontend/src/app/coworking/widgets/room-dialogue/room-dialogue.component.spec.ts]: Removed

File: frontend/src/app/coworking/room-reservation/reservation-table.service.ts (hunk 2 above)
Comment (tmattvu) [reply frontend/src/app/coworking/room-reservation/reservation-table.service.ts:99]: Looping in @aarjavjain2002

File: frontend/src/app/coworking/coworking.models.ts (hunk 1 above)
Comment (tmattvu) [reply frontend/src/app/coworking/coworking.models.ts:52]: Looping in @aarjavjain2002

File: backend/services/coworking/reservation.py [hunk 15]
```diff
@@ -348,11 +348,13 @@ def get_map_reserved_times_by_date(
...
-                    # TODO: If making group reservations, need to change this.
-                    if reservation.users[0].id == subject.id:
-                        time_slots_for_room[idx] = RoomState.SUBJECT_RESERVED.value
-                    else:
+                    flag = True
+                    for user in reservation.users:
+                        if user.id == subject.id:
+                            time_slots_for_room[idx] = RoomState.SUBJECT_RESERVED.value
+                            flag = True
+                            break
```
[the remark this answers is not in the document]
Comment (tmattvu) [reply backend/services/coworking/reservation.py:356]: Looping in @aarjavjain2002

**Comment by ajaygandecha** [comment]
@tmattvu / @aarjavjain2002 / @yuvrajjain2003 what is the progress here? I am working on cleaning up branches and PRs so I wanted to check if this needs re-review.

**Comment by aarjavjain2002** [comment]
> @tmattvu / @aarjavjain2002 / @yuvrajjain2003 what is the progress here? I am working on cleaning up branches and PRs so I wanted to check if this needs re-review.

Hi Ajay. I believe I had cleared up all issues that were pertaining to room reservations and updated Matt on the same. I am not sure if he has any more work left on the branch. @tmattvu all yours from here!

**Comment by tmattvu** [comment]
> @tmattvu / @aarjavjain2002 / @yuvrajjain2003 what is the progress here? I am working on cleaning up branches and PRs so I wanted to check if this needs re-review.

Hey Ajay, sorry for the late response, but I believe all I have to do left is testing. I've been a bit busy since the semester ended but I'll try to tackle it later this week!

**Comment by ajaygandecha** [comment]
@aarjavjain2002 @yuvrajjain2003 @tmattvu I am going to close this PR, since production has diverged extremely far from the last commit on this PR. I recommend attempting to merge `main` into this and resolve changes!

In addition, for this PR to be approved, I recommend a bit of a re-design to conform to the new Material 3 standards. Here are a few things I would recommend:

* You can greatly simplify the group reservation frontend to just include a single autocompleting input with user chips. The `<user-chip-list>` widget was created by @jadekeegan to achieve this, and I have used it in many places in the new site. I recommend checking out this use case here in the relocated section admin page:

  https://github.com/unc-csxl/csxl.unc.edu/blob/82885ede514faec99c2a93df34c15c3a1dd3336c/frontend/src/app/academics/academics-admin/section/section-editor/section-editor.component.html#L62-L65
  
  This looks like the following on the frontend:
  
  Screenshot 2024-07-13 at 1 15 12 AM
  
  (Profile images will show up, if applicable)
  
  In the future, assuming the empty box below is replaced with the interactive seating chart map, this is the aspirational M3 design of the coworking drop-in screen based on the Figma:
  

  Screenshot 2024-07-13 at 1 16 51 AM

* Revert changes to rooms, as this has already been accounted for in the M3 update (I believe).

* There will be a better guide for updating old developers to the new conventions with the summer site upgrade coming soon, so feel free to ask any questions if you have them!
